varats-core/varats/paper/paper_config.py

```diff
@@ -256,9 +259,17 @@ def __init__(
         self.__project_name = project_name
 
     def versions(self) -> tp.List[bb.source.base.Variant]:
+        cache_path = self.fetch()
+        git_rev_list = git['rev-list', '--abbrev-commit', '--abbrev=10',
+                           '--all']
+        rev_list: tp.List[str] = []
+        with pb.local.cwd(cache_path):
+            rev_list = list(git_rev_list(self.refspec).strip().split('\n'))
+
+        rev_list = list(filter(self.version_filter, rev_list))
+        rev_list = rev_list[:self.limit] if self.limit else rev_list
+        variants = [base.Variant(version=rev, owner=self) for rev in rev_list]
```

Comment:
You could avoid constructing variants that will get filtered out anyways by merging this line with the return statement:

```suggestion
        return [base.Variant(version=rev, owner=self) for rev in rev_list if proj_filter(rev)]
```